src/coreclr/vm/codeman.cpp

```diff
@@ -3917,126 +3984,380 @@ StubCodeBlockKind EEJitManager::GetStubCodeBlockKind(RangeSection * pRangeSectio
 
-    // Find the nibble with the header in the DWORD
+            size_t dwordIndex = GetDwordIndex(delta);
+            size_t nibbleIndex = GetNibbleIndex(delta);
 
-    while (startPos && !(tmp & NIBBLE_MASK))
-    {
-        tmp = tmp >> NIBBLE_SIZE;
-        startPos--;
-    }
+            PTR_DWORD pMap = pHp->pHdrMap;
 
-    if (startPos == 0 && tmp == 0)
-        return 0;
+            DWORD dword;
+            DWORD nibble;
+
+            DWORD  offset   = ADDR2OFFS(delta); // this is the offset inside the bucket + 1
+            _ASSERTE(offset == (offset & NIBBLE_MASK));
+
+            // #1 look up DWORD represnting current PC
+            PREFIX_ASSUME(pMap + dwordIndex != NULL);
+            dword = VolatileLoadWithoutBarrier<DWORD>(pMap + dwordIndex);
+
+            // if DWORD is empty, we can skip steps 2,3, and 4
+            if (dword)
+            {
+                // #2 if DWORD is a pointer, then we can return
+                if (IsPointer(dword))
+                {
+                    TADDR newAddr = base + DecodePointer(dword);
+                    return newAddr;
+                }
+
+                // #3 if DWORD is nibbles and corresponding nibble is intialized and points to an equal or earlier address, return the corresponding address
+                nibble = GetNibble(dword, nibbleIndex);
+                if ((nibble) && (nibble <= offset) )
+                {
+                    return base + NibbleToRelativeAddress(dwordIndex, nibbleIndex, nibble);
+                }
+
+                // #4 find preceeding nibble and return if found
+                // if nibbleIndex == 0, then there are no preceeding nibbles in this DWORD
+                if(nibbleIndex != 0)
+                {
+                    DWORD preceedingNibbleMask = ~0x0u << (32 - nibbleIndex * 4);
+                    if(dword & preceedingNibbleMask)
+                    {
+                        size_t ctz = NibbleBitScanForward(dword & preceedingNibbleMask);
+                        size_t nibbleToCheck = (31 - ctz) / 4;
+                        nibble = GetNibble(dword, nibbleToCheck);
+                        if(nibble) 
+                        {
+                            return base + NibbleToRelativeAddress(dwordIndex, nibbleToCheck, nibble);
+                        }
+                    }
+                }
+            }
+
+            // #5.1 read previous DWORD. If no such DWORD return 0.
+            if (dwordIndex == 0)
+            {
+                return 0;
+            }
+            dwordIndex--;
+            nibbleIndex = NIBBLES_PER_DWORD;
+
+            PREFIX_ASSUME(pMap + dwordIndex != NULL);
+            dword = VolatileLoadWithoutBarrier<DWORD>(pMap + dwordIndex);
+
+            if(dword)
+            {
+                // #5.2 if DWORD is a pointer, then we can return
+                if (IsPointer(dword))
+                {
+                    return base + DecodePointer(dword);
+                }
```

Comment:
Same question as case 2 - don't you need to read the nibble that the pointer points to inorder to check that the method isn't being deleted at the same time as you're looking it up?